pkg/xds/cache/once/cache.go

```diff
@@ -0,0 +1,58 @@
+package once
+
+import (
+	"context"
+	"time"
+
+	"github.com/patrickmn/go-cache"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/kumahq/kuma/pkg/metrics"
+)
+
+type Cache struct {
+	cache   *cache.Cache
+	onceMap *omap
+	metrics *prometheus.GaugeVec
+}
+
+func New(expirationTime time.Duration, name string, metrics metrics.Metrics) (*Cache, error) {
+	metric := prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: name,
+		Help: "Summary of " + name,
+	}, []string{"operation", "result"})
+	if err := metrics.Register(metric); err != nil {
+		return nil, err
+	}
+	return &Cache{
+		cache:   cache.New(expirationTime, time.Duration(int64(float64(expirationTime)*0.9))),
+		onceMap: newMap(),
+		metrics: metric,
+	}, nil
+}
+
+func (c *Cache) Get(ctx context.Context, key string, fn func(context.Context, string) (interface{}, error)) (interface{}, error) {
```

Comment:
Yeah, there was duplication indeed, but it's quite problematic to introduce a well-understandable and reusable abstraction here. I don't mind going with this `once.Cache` abstraction, but we have to do something with the interface because it's really unclear what `Get` does. 

Maybe we can introduce new types:
```go
type Key string
type Value interface{}
type Refresher func(context.Context, Key) (Value, error)
```
and rename `Get` to `GetOrRefresh` so it will look like:

```go
func (c *Cache) GetOrRefresh(ctx context.Context, key Key, refresher Refresher) (Value, error)
```
and document this method with a comment something like:

```go
// GetOrRefresh method returns cached Value if it was created within the last `expirationTime` time period.
// If the Value is older than `expirationTime` then Refresher will be launched and will update the Value.
// Regardless of the number of goroutines that have run into the expired Value, Refresher will be called only
// once, all goroutines will be blocked until the new Value returned by Refresher.
func (c *Cache) GetOrRefresh(ctx context.Context, k Key, r Refresher) (Value, error) {
```

WDYT?

Reply:
Sounds good!
How about `GetOrCompute()` as Refresh sounds like it will only work when the item has been in the cache?

For the `Key` what's the motivation for adding the subtype?

Reply:
The main reason for questioning the introduction of subtypes is that the underlying lib [`go-cache`](https://github.com/patrickmn/go-cache) uses `string` as key and `interface{}` as value.

Reply:
In the end went for `GetOrRetrieve` and didn't add subtypes for `key` and `value`. Let me know what you think.

Reply:
I like `GetOrRetrieve` 👍 

Subtypes help to make the signature more readable. If you define it like:
```go
type FetcherFunc func(context.Context, string) (interface{}, error)
```
then it's not clear that `FetcherFunc` takes `key` as a parameter and returns `value`.

Also, `interface{}` is used twice - as a return value of `GetOrRetrieve` and as a return value of `FetcherFunc`. I think it makes sense to show that those `interface{}` are the same type. For example, if in the future we would like to introduce a new type for `Value`.

But I don't mind going with just `string` and `interface{}`